src/validation.cpp

```diff
@@ -1107,6 +1085,12 @@ bool MemPoolAccept::ReplacementChecks(Workspace& ws)
     for (auto it : all_conflicts) {
         m_subpackage.m_changeset->StageRemoval(it);
     }
+
+    if (const auto err_string{ImprovesFeerateDiagram(*m_subpackage.m_changeset)}) {
+        // If we can't calculate a feerate, it's because the cluster size limits were hit.
+        return state.Invalid(TxValidationResult::TX_RECONSIDERABLE, "replacement-failed", err_string->second);
```

Comment:
e6315c24326: In both single and package settings, `ImprovesFeerateDiagram` is where cluster size limits are checked for the first time if RBF is happening. I think this is fine since the check is cached and many RBF checks are cheaper than `CheckMemPoolPolicyLimits`. But I don't think exceeding cluster limits should be considered `TX_RECONSIDERABLE`. It should instead be a `TX_MEMPOOL_POLICY` error. Also, I think the error string should be "too-large-cluster" (but that's more of a nit).

Putting `CheckMemPoolPolicyLimits` before replacement checks would automatically fix this.
We could also examine the `err_string` here and assign `TX_RECONSIDERABLE` if it's actually a feerate diagram error, `TX_MEMPOOL_POLICY` if it's a cluster limit error.

Reply:
> I also considered having all cluster size limits being treated as RECONSIDERABLE, but decided that is also not great, partly because we certainly shouldn't let a transaction whose cluster size just with its own unconfirmed ancestors be treated as RECONSIDERABLE

👍 definitely think we should not allow reconsideration in that case

Sounds good to me to make it non-reconsiderable for now, and maybe change it in the future in a consistent way. I agree it's possible for a child to make a tx no longer break cluster limits, but it's not clear to me how important it is to support that case.